dvc/utils/strictyaml.py

```diff
@@ -206,12 +230,25 @@ def __init__(
             message += f": {len(self.exc.errors)} errors"
         super().__init__(f"{message}")
 
-    def _prepare_context(self, data: typing.Mapping) -> List[object]:
+    def _prepare_context(self, data: Mapping) -> List[object]:
         lines: List[object] = []
         for index, error in enumerate(self.exc.errors):
             if index and lines[-1]:
                 lines.append("")
-            line, col, step = determine_linecol(data, error.path)
+
+            try:
+                line, col = determine_linecol(
+                    data,
+                    error.path,
+                    # Handle the case where a validation error indicates that additional
+                    # keys are not permitted.
+                    key_or_value="extra keys not allowed" in error.error_message,
+                )
+            except KeyError:
+                # Handle the case where a validation error occurs because a required
+                # key is missing.
+                line, col = determine_linecol(data, error.path[:-1], True)
```

Comment:
I don't want this to be coupled with dvcfile. We should just move up the path if we don't find the keys.